### PR TITLE
[WIP] Unit tests for E3-E7

### PR DESCRIPTION
**Pull request by SamuilsRulovs**
Added Unit test for:

basin-faceted-brep
basin-tessellation
beam-curved-i-shape-tessellated
beam-straight-i-shape-tessellated
column-straight-rectangle-tessellation
slab-tessellated-unique-vertices
mapped-shape-with-multiple-items
mapped-shape-without-transformation
mapped-shape-with-transformation

Edited Unit test for: 

tessellated-item

To do Unit test for:

curve.ifc #232

### CODE REVIEW

**Review by pjanck — CHANGES_REQUESTED** [review]
The pictures of `basin-faceted-brep` need to be corrected. Otherwise a good start.

**Comment by SamuilsRulovs** [comment]
I'm not sure, if I use correct Entities in Test "IFCHasAnEssentialEntity".  Could you please check this?

File: UnitTests/Schemas/IFC4X1/beam-curved-i-shape-tessellated/src/beam-curved-i-shape-tessellated.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+};
+
+TEST_F(BeamCurvedIShapeTessellated, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(27));
+}
+
+TEST_F(BeamCurvedIShapeTessellated, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCBEAM"; });
+	EXPECT_NE(result, express_model->entities.end());
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/beam-curved-i-shape-tessellated/src/beam-curved-i-shape-tessellated.cpp]: Please also add tests for these entities:
- IFCTRIANGULATEDFACESET

File: UnitTests/Schemas/IFC4X1/beam-straight-i-shape-tessellated/src/beam-straight-i-shape-tessellated.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+};
+
+TEST_F(BeamStraightIShapeTessellated, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(20));
+}
+
+TEST_F(BeamStraightIShapeTessellated, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCBEAM"; });
+	EXPECT_NE(result, express_model->entities.end());
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/beam-straight-i-shape-tessellated/src/beam-straight-i-shape-tessellated.cpp]: Please also add tests for these entities:
- IFCTRIANGULATEDFACESET

File: UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+};
+
+TEST_F(ColumnStraightRectangleTessellation, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(26));
+}
+
+TEST_F(ColumnStraightRectangleTessellation, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCRELCONTAINEDINSPATIALSTRUCTURE"; });
+	EXPECT_NE(result, express_model->entities.end());
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp]: Please also add tests for these entities:
- IFCTRIANGULATEDFACESET

Remove the currently tested entity.

File: UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+TEST_F(ColumnStraightRectangleTessellation, ImageIsSaved)
+{
+	// Arrange
+	buw::Image4b image = renderer->captureImage();
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp:76]: The column seems to be very far away from the viewing point. This should be examined.

File: UnitTests/Schemas/IFC4X1/mapped-shape-with-multiple-items/src/mapped-shape-with-multiple-items.cpp [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+};
+
+TEST_F(MappedShapeWithMultipleItems, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(50));
+}
+
+TEST_F(MappedShapeWithMultipleItems, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCMAPPEDITEM"; });
+	EXPECT_NE(result, express_model->entities.end());
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/mapped-shape-with-multiple-items/src/mapped-shape-with-multiple-items.cpp]: Please also add tests for these entities:
- IFCEXTRUDEDAREASOLID
- IFCRECTANGLEPROFILEDEF
- IFCCARTESIANTRANSFORMATIONOPERATOR3DNONUNIFORM

Also, please use `std::count_if` in a similar way, to check that there are 4 IfcMappedItem-s as well as 4 IfcCartesianTransf...-s. This hsould be a new `TEST_F`.

File: UnitTests/Schemas/IFC4X1/mapped-shape-with-transformation/src/mapped-shape-with-transformation.cpp [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+};
+
+TEST_F(MappedShapeWithTransformation, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(40));
+}
+
+TEST_F(MappedShapeWithTransformation, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCMAPPEDITEM"; });
+	EXPECT_NE(result, express_model->entities.end());
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/mapped-shape-with-transformation/src/mapped-shape-with-transformation.cpp]: Please also add tests for these entities:

- IFCEXTRUDEDAREASOLID
- IFCRECTANGLEPROFILEDEF
- IFCCARTESIANTRANSFORMATIONOPERATOR3DNONUNIFORM

File: UnitTests/Schemas/IFC4X1/mapped-shape-without-transformation/src/mapped-shape-without-transformation.cpp [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+};
+
+TEST_F(MappedShapeWithoutTransformation, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(38));
+}
+
+TEST_F(MappedShapeWithoutTransformation, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCMAPPEDITEM"; });
+	EXPECT_NE(result, express_model->entities.end());
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/mapped-shape-without-transformation/src/mapped-shape-without-transformation.cpp]: Please also add tests for these entities:
- IFCEXTRUDEDAREASOLID
- IFCRECTANGLEPROFILEDEF

File: UnitTests/Schemas/IFC4X1/slab-tessellated-unique-vertices/src/slab-tessellated-unique-vertices.cpp [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+};
+
+TEST_F(SlabTessellatedUniqueVertices, AllEntitiesAreRead) {
+	EXPECT_THAT(express_model->entities.size(), Eq(26));
+}
+
+TEST_F(SlabTessellatedUniqueVertices, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCSLAB"; });
+	EXPECT_NE(result, express_model->entities.end());
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/slab-tessellated-unique-vertices/src/slab-tessellated-unique-vertices.cpp]: This should check for `IFCTRIANGULATEDFACESET`

File: testdata/IFC4x1ExampleFiles/official/curve.ifc [hunk 9]
```diff
@@ -0,0 +1,37 @@
+ISO-10303-21;
```
Comment (pjanck) [inline testdata/IFC4x1ExampleFiles/official/curve.ifc]: This is not an "official" file, move it one level up to the IFC4x1ExampleFiles folder.

**Review by pjanck — CHANGES_REQUESTED** [review]

File: UnitTests/Schemas/IFC4X1/beam-curved-i-shape-tessellated/src/beam-curved-i-shape-tessellated.cpp (hunk 1 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/beam-curved-i-shape-tessellated/src/beam-curved-i-shape-tessellated.cpp]: Resolved: afcbbd4

File: UnitTests/Schemas/IFC4X1/beam-straight-i-shape-tessellated/src/beam-straight-i-shape-tessellated.cpp (hunk 2 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/beam-straight-i-shape-tessellated/src/beam-straight-i-shape-tessellated.cpp]: Resolved: afcbbd4

File: UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp (hunk 3 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp]: Resolved: afcbbd4

File: UnitTests/Schemas/IFC4X1/mapped-shape-with-multiple-items/src/mapped-shape-with-multiple-items.cpp (hunk 5 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/mapped-shape-with-multiple-items/src/mapped-shape-with-multiple-items.cpp]: Resolved: afcbbd4

File: UnitTests/Schemas/IFC4X1/mapped-shape-with-transformation/src/mapped-shape-with-transformation.cpp (hunk 6 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/mapped-shape-with-transformation/src/mapped-shape-with-transformation.cpp]: Resolved: afcbbd4

File: UnitTests/Schemas/IFC4X1/mapped-shape-without-transformation/src/mapped-shape-without-transformation.cpp (hunk 7 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/mapped-shape-without-transformation/src/mapped-shape-without-transformation.cpp]: Resolved: afcbbd4

File: UnitTests/Schemas/IFC4X1/slab-tessellated-unique-vertices/src/slab-tessellated-unique-vertices.cpp (hunk 8 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/slab-tessellated-unique-vertices/src/slab-tessellated-unique-vertices.cpp]: Resolved: afcbbd4

File: testdata/IFC4x1ExampleFiles/official/curve.ifc (hunk 9 above)
Comment (SamuilsRulovs) [reply testdata/IFC4x1ExampleFiles/official/curve.ifc]: Resolved: afcbbd4

**Review by pjanck — APPROVED** [review]

File: UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp (hunk 4 above)
Comment (pjanck) [reply UnitTests/Schemas/IFC4X1/column-straight-rectangle-tessellation/src/column-straight-rectangle-tessellation.cpp:76]: Please submit this as a new issue - this should be looked at, why the extent of this model is somehow falsely estimated.

**Pull request merged**